### PR TITLE
Added a method to compute the Tutte Symmetric function of a graph

### PR DESCRIPTION
**Pull request by s-s-sawant**
Added a method that returns the Tutte symmetric function of the graph. Added references accordingly.

Tutte symmetric function of a graph

For a graph G (may contain multiedges and loops), the new methods returns the Tutte symmetric function of the graph. It is implemented by using the expansion of the Tutte symmetric function in the monomial symmetric function basis.
The references are added and the index.rst file is updated accordingly.

The Tutte symmetric function of a graph is a chromatic invariant that generalizes the Tutte polynomial to a symmetric function. One can recover the chromatic symmetric function of a graph from its Tutte symmetric function. Analogous to the Stanley's tree isomorphism conjecture, the graph classes distinguished by their Tutte symmetric function are actively investigated.
The newly added method returns the monomial symmetric function expansion of the Tutte symmetric function of a graph.

### :memo: Checklist

- [x] The title is concise and informative.
- [x] The description explains in detail what this PR is about.
- [ ] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation and checked the documentation preview.

### :hourglass: Dependencies
No pending dependencies

### Comment:
All tests are passed. However, unable to run docbuild. Shows the following error:
"File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code"

### CODE REVIEW

**Comment by sagar-math** [comment]
For a graph G, its Tutte symmetric function is a chromatic invariant that generalizes the Tutte polynomial to a symmetric function. 
We add a method to compute the Tutte symmetric function of a graph using its expansion in the augmented monomial symmetric function basis.

Analogous to the Stanley's tree isomorphism conjecture, determining graph classes that are distinguished by their Tutte symmetric function is actively investigated.

@tscrim @amritanshu-prasad @mantepse

File: src/sage/graphs/graph.py [hunk 1]
```diff
@@ -4122,6 +4122,90 @@ def asc(sigma):
...
+        m = SymmetricFunctions(R).m()
+        ret = m.zero()
+        V = self.vertices()
+        E = list(self.edges(labels=False))
+
+        def mono(pi):
+            arcs = 0
+            for s in pi:
+                for u in s:
+                    arcs += sum(E.count((u, v)) for v in s if self.has_edge(u, v))
```
Comment (dcoudert) [inline src/sage/graphs/graph.py]: the call to `E.count((u, v))` takes time $O(m)$. You can do much better by pre-computing the multiplicity of the edges using for instance:
```
from collections import Counter
M = Counter(E)
```
and then use
```
arcs += sum(M[(u, v)] for v in s if self.has_edge(u, v))
```
this way, you get the multiplicity of `(u, v)` in time $O(1)$.

File: src/sage/graphs/graph.py [hunk 2]
```diff
@@ -4122,6 +4122,90 @@ def asc(sigma):
...
+            return arcs
+
+        for pi in SetPartitions(V):
+            ret += prod([factorial(i) for i in pi.to_partition().to_exp()])*m[pi.to_partition()]*(1+t)**mono(pi)
```
Comment (dcoudert) [inline src/sage/graphs/graph.py]: you should use an extra variable to store `pi.to_partition()`. You may also precompute and store the values of `factorial(i)` if you know the largest value of `I`.

**Review by dcoudert — CHANGES_REQUESTED** [review]
further improvements are certainly possible.

File: src/sage/graphs/graph.py [hunk 3]
```diff
@@ -4122,6 +4122,95 @@ def asc(sigma):
             ret += M.term(sigma.to_composition(), t**asc(sigma))
         return ret
 
+    @doc_index("Coloring")
+    def tutte_symmetric_function(self, R=None, t=None):
+        r"""
+        Return the Tutte symmetric function of ``self``.
+
+        Let `G` be a graph. The Tutte symmetric function `XB_G` of the graph `G` was introduced in [Sta1998]_. We present
```
Comment (dcoudert) [inline src/sage/graphs/graph.py]: align on 80 columns mode

File: src/sage/graphs/graph.py [hunk 4]
```diff
@@ -4122,6 +4122,95 @@ def asc(sigma):
...
+        the equivalent definition given in [CS2022]_.
+
+        .. MATH::
+
+            XB_G = \sum_{\pi \vdash V} (1+t)^{e(\pi)} \tilde{m}_{\lambda(\pi)},
+
+        where the sum ranges over all set-partitions `\pi` of the vertex set `V`, `\lambda(pi)` is the
```
Comment (dcoudert) [inline src/sage/graphs/graph.py]: 80 columns mode

File: src/sage/graphs/graph.py [hunk 5]
```diff
@@ -4122,6 +4122,95 @@ def asc(sigma):
...
+        partition determined by the sizes of the blocks of `\pi`, and `e(pi)` is the number of edges whose endpoints
+        lie in the same block of `\pi`. In particular, the coefficients of `XB_G`
+        when expanded in terms of augmented monomial symmetric functions are polynomials in `t` with
+        non-negative integer coefficients.
+
+        For an integer partition `\lambda = 1^{r_1}2^{r_2}\cdots` expressed in the exponential notation, the augmented monomial symmetric function is defined as
```
Comment (dcoudert) [inline src/sage/graphs/graph.py]: 80 columns mode

File: src/sage/graphs/graph.py [hunk 6]
```diff
@@ -4122,6 +4122,95 @@ def asc(sigma):
...
+
+        .. MATH::
+
+            \tilde{m}_{\lambda} = \(\prod_{i} r_i!\) m_{\lambda}.
+
+
```
Comment (dcoudert) [inline src/sage/graphs/graph.py:4152]: 1 empty line should be enough

File: src/sage/graphs/graph.py [hunk 7]
```diff
@@ -4122,6 +4122,95 @@ def asc(sigma):
...
+        - ``t`` -- (optional) the parameter `t`; uses the variable `t` in
+            `\ZZ[t]` by default
+
+        EXAMPLES:
+
+            sage: p = SymmetricFunctions(ZZ).p()                                        # needs sage.combinat sage.modules
+            sage: G = Graph([[1,2],[2,3],[3,4],[4,1],[1,3]])
+            sage: XB_G = G.tutte_symmetric_function(); XB_G                             # needs sage.combinat sage.modules
+            24*m[1, 1, 1, 1] + (10*t+12)*m[2, 1, 1] + (4*t^2+10*t+6)*m[2, 2] + (2*t^3+8*t^2+10*t+4)*m[3, 1] + (t^5+5*t^4+10*t^3+10*t^2+5*t+1)*m[4]
```
Comment (dcoudert) [inline src/sage/graphs/graph.py]: you may display the result on 2 lines to avoid very long lines

File: src/sage/graphs/graph.py [hunk 8]
```diff
@@ -4122,6 +4122,95 @@ def asc(sigma):
...
+
+        if t is None:
+            t = ZZ['t'].gen()
+        if R is None:
+            R = t.parent()
+        m = SymmetricFunctions(R).m()
+        ret = m.zero()
+        V = self.vertices()
+        E = list(self.edges(labels=False))
+        M = Counter(E)
```
Comment (dcoudert) [inline src/sage/graphs/graph.py]: Here you can certainly do directly `M = Counter(self.edge_iterator(labels=False))`, thus avoiding intermediate variable `E` which is not used elsewhere.

File: src/sage/graphs/graph.py (hunk 3 above)
Comment (s-s-sawant) [reply src/sage/graphs/graph.py]: corrected

File: src/sage/graphs/graph.py (hunk 4 above)
Comment (s-s-sawant) [reply src/sage/graphs/graph.py]: corrected

File: src/sage/graphs/graph.py (hunk 5 above)
Comment (s-s-sawant) [reply src/sage/graphs/graph.py]: corrected

File: src/sage/graphs/graph.py (hunk 6 above)
Comment (s-s-sawant) [reply src/sage/graphs/graph.py:4152]: corrected

File: src/sage/graphs/graph.py (hunk 8 above)
Comment (s-s-sawant) [reply src/sage/graphs/graph.py]: corrected

File: src/sage/graphs/graph.py (hunk 1 above)
Comment (s-s-sawant) [reply src/sage/graphs/graph.py]: Corrected.

File: src/sage/graphs/graph.py [hunk 9]
```diff
@@ -4122,6 +4122,103 @@ def asc(sigma):
...
+
+            \tilde{m}_{\lambda} = \(\prod_{i} r_i!\) m_{\lambda}.
+
+        INPUT:
+
+        - ``R`` -- (optional) the base ring for the symmetric functions;
+            this uses `\ZZ` by default
+
+        - ``t`` -- (optional) the parameter `t`; uses the variable `t` in
+            `\ZZ[t]` by default
```
Comment (tscrim) [inline src/sage/graphs/graph.py]:
```suggestion
        - ``R`` -- (default: the parent of ``t``) the base ring for the symmetric functions
        - ``t`` -- (default: `t` in `\ZZ[t]`) the parameter `t`
```

File: src/sage/graphs/graph.py [hunk 10]
```diff
@@ -4122,6 +4122,103 @@ def asc(sigma):
...
+
+        EXAMPLES:
```
Comment (tscrim) [inline src/sage/graphs/graph.py]:
```suggestion
        EXAMPLES::
```

File: src/sage/graphs/graph.py [hunk 11]
```diff
@@ -4122,6 +4122,103 @@ def asc(sigma):
...
+            sage: p = SymmetricFunctions(ZZ).p()                                        # needs sage.combinat sage.modules
+            sage: G = Graph([[1,2],[2,3],[3,4],[4,1],[1,3]])
+            sage: XB_G = G.tutte_symmetric_function(); XB_G                             # needs sage.combinat sage.modules
+            24*m[1, 1, 1, 1] + (10*t+12)*m[2, 1, 1] + (4*t^2+10*t+6)*m[2, 2]
+            + (2*t^3+8*t^2+10*t+4)*m[3, 1]
+            + (t^5+5*t^4+10*t^3+10*t^2+5*t+1)*m[4]
+            sage: p(XB_G)                                                               # needs sage.combinat sage.modules
+            p[1, 1, 1, 1] + 5*t*p[2, 1, 1] + 2*t^2*p[2, 2]
+            + (2*t^3+8*t^2)*p[3, 1] + (t^5+5*t^4+8*t^3)*p[4]
+
```
Comment (tscrim) [inline src/sage/graphs/graph.py]:
```suggestion
            24*m[1, 1, 1, 1] + (10*t+12)*m[2, 1, 1] + (4*t^2+10*t+6)*m[2, 2]
             + (2*t^3+8*t^2+10*t+4)*m[3, 1]
             + (t^5+5*t^4+10*t^3+10*t^2+5*t+1)*m[4]
            sage: p(XB_G)                                                               # needs sage.combinat sage.modules
            p[1, 1, 1, 1] + 5*t*p[2, 1, 1] + 2*t^2*p[2, 2]
             + (2*t^3+8*t^2)*p[3, 1] + (t^5+5*t^4+8*t^3)*p[4]

```

File: src/sage/graphs/graph.py [hunk 12]
```diff
@@ -4122,6 +4122,103 @@ def asc(sigma):
...
+        Graphs are allowed to have multiedges and loops::
+
+            sage: G = Graph([[1,2],[2,3],[2,3]], multiedges = True)
+            sage: XB_G = G.tutte_symmetric_function(); XB_G                             # needs sage.combinat sage.modules
+            6*m[1, 1, 1] + (t^2+3*t+3)*m[2, 1] + (t^3+3*t^2+3*t+1)*m[3]
+
+
+        We check that at `t = -1`, we recover the usual chromatic symmetric
```
Comment (tscrim) [inline src/sage/graphs/graph.py]:
```suggestion
            6*m[1, 1, 1] + (t^2+3*t+3)*m[2, 1] + (t^3+3*t^2+3*t+1)*m[3]

        We check that at `t = -1`, we recover the usual chromatic symmetric
```

File: src/sage/graphs/graph.py [hunk 13]
```diff
@@ -4122,6 +4122,103 @@ def asc(sigma):
...
+            sage: XB_G = G.tutte_symmetric_function(t=-1); XB_G                         # needs sage.combinat sage.modules
+            120*m[1, 1, 1, 1, 1] + 36*m[2, 1, 1, 1] + 12*m[2, 2, 1]
+            + 2*m[3, 1, 1] + m[3, 2]
+            sage: X_G = G.chromatic_symmetric_function(); X_G                           # needs sage.combinat sage.modules
+            p[1, 1, 1, 1, 1] - 4*p[2, 1, 1, 1] + 3*p[2, 2, 1] + 3*p[3, 1, 1]
+            - 2*p[3, 2] - 2*p[4, 1] + p[5]
+            sage: XB_G == X_G                                                           # needs sage.combinat sage.modules
+            True
+
+        """
```
Comment (tscrim) [inline src/sage/graphs/graph.py]:
```suggestion
            sage: XB_G == X_G                                                           # needs sage.combinat sage.modules
            True
        """
```
(I am trying to uniformize this.)

File: src/sage/graphs/graph.py (hunk 8 above)
Comment (tscrim) [reply src/sage/graphs/graph.py]: This is not changed in what I can see.

File: src/sage/graphs/graph.py [hunk 14]
```diff
@@ -4122,6 +4122,103 @@ def asc(sigma):
...
+        def mono(pi):
+            arcs = 0
+            for s in pi:
+                for u in s:
+                    arcs += sum(M[(u, v)] for v in s if self.has_edge(u, v))
+            return arcs
+
+        for pi in SetPartitions(V):
+            pa = pi.to_partition()
+            ret += prod([fact[i] for i in pa.to_exp()])*m[pa]*(1+t)**mono(pi)
```
Comment (tscrim) [inline src/sage/graphs/graph.py]:
```suggestion
            ret += prod(fact[i] for i in pa.to_exp()) * m[pa] * (1+t)**mono(pi)
```
PEP8 spacing and removing an unnecessary list construction.

File: src/sage/graphs/graph.py [hunk 15]
```diff
@@ -4122,6 +4122,103 @@ def asc(sigma):
...
+
+        if t is None:
+            t = ZZ['t'].gen()
+        if R is None:
+            R = t.parent()
+        m = SymmetricFunctions(R).m()
+        ret = m.zero()
+        V = self.vertices()
+        M = Counter(list(self.edges(labels=False)))
+        fact = [factorial(i) for i in range(len(V)+1)]
```
Comment (tscrim) [inline src/sage/graphs/graph.py]:
```suggestion
        fact = [1]
        fact.extend(fact[-1] * i for i in range(1, len(V)+1))
```
This is a fun use of the `extend` method; it is slightly complicated, but it should be faster due to fewer multiplications. Of course, this is definitely a micro-optimization...

**Review by tscrim — COMMENTED** [review]
A few other little things from me.

File: src/sage/graphs/graph.py (hunk 15 above)
Comment (s-s-sawant) [reply src/sage/graphs/graph.py]: That does save some repeated computations

Comment (s-s-sawant) [reply src/sage/graphs/graph.py]: corrected

**Comment by dcoudert** [comment]
@tscrim, do you know why the CI is not running for this PR ? it would be useful to finalize the review. I think the code is ok now.

**Comment by mantepse** [comment]
> @tscrim, do you know why the CI is not running for this PR ? it would be useful to finalize the review. I think the code is ok now.

because the author is a first time contributor.

File: src/sage/graphs/graph.py [hunk 16]
```diff
@@ -4122,6 +4122,101 @@ def asc(sigma):
...
+        the exponential notation, the augmented monomial symmetric function
+        is defined as
+
+        .. MATH::
+
+            \tilde{m}_{\lambda} = \(\prod_{i} r_i!\) m_{\lambda}.
+
+        INPUT:
+
+        - ``R`` -- (default: the parent of ``t``) the base ring for the symmetric
```
Comment (dcoudert) [inline src/sage/graphs/graph.py]: there is a trailing white space at the end of the line. remove it.

**Review by dcoudert — APPROVED** [review]
LGTM.

**Comment by dcoudert** [comment]
What's the purpose of this last commit ? why is method `matching` in this PR ?

**Comment by s-s-sawant** [comment]
> What's the purpose of this last commit ? why is method `matching` in this P

A conflict was raised highlighting `tutte_symmetric_function`, `matching`  and `is_factor_critical`. The new commit was made while resolving the conflict.

**Comment by dcoudert** [comment]
Methods `matching` and `is_factor_critical` have been moved to `src/sage/graphs/matching.py` in #38218. 
So you should not put them back to `src/sage/graphs/graph.py`.

File: src/sage/graphs/graph.py [hunk 17]
```diff
@@ -4122,6 +4122,101 @@ def asc(sigma):
...
+            sage: X_G = G.chromatic_symmetric_function(); X_G                           # needs sage.combinat sage.modules
+            p[1, 1, 1, 1, 1] - 4*p[2, 1, 1, 1] + 3*p[2, 2, 1] + 3*p[3, 1, 1]
+             - 2*p[3, 2] - 2*p[4, 1] + p[5]
+            sage: XB_G == X_G                                                           # needs sage.combinat sage.modules
+            True
+        """
+        from sage.combinat.sf.sf import SymmetricFunctions
+        from sage.combinat.set_partition import SetPartitions
+        from sage.misc.misc_c import prod
+        from sage.arith.misc import factorial
```
Comment (dcoudert) [inline src/sage/graphs/graph.py]: you don't need to import factorial anymore

File: src/sage/graphs/graph.py [hunk 18]
```diff
@@ -4122,6 +4122,101 @@ def asc(sigma):
...
+        from collections import Counter
+
+        if t is None:
+            t = ZZ['t'].gen()
+        if R is None:
+            R = t.parent()
+        m = SymmetricFunctions(R).m()
+        ret = m.zero()
+        V = self.vertices()
+        M = Counter(list(self.edges(labels=False)))
```
Comment (tscrim) [inline src/sage/graphs/graph.py]: Does this really need to get the list of edges? I thought having an iterator is sufficient.In particular, this does not match @dcoudert 's suggestion.

**Review by dcoudert — APPROVED** [review]
LGTM.

**Pull request merged**